<h1 dir="rtl" style="font-family: Vazir; width: 85%;">تمرین شماره ۱: درس پردازش زبان طبیعی - دانشگاه تهران، پائیز ۱۴۰۳</h1>

<div dir="rtl" style="font-family: Vazir; width: 85%; font-size: 18px;">
نام: عرفان شهابی
<br/>


<div dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
سوالات خودتان را می‌توانید از طریق ایمیل
<code>ghanizadeh.amin@ut.ac.ir</code>
 از طراح تمرین اول بپرسید.

<div dir="rtl" style="font-family: Vazir; width: 85%; font-size: 18px; color: red; font-weight: bold;">
قوانین و توضیحاتی آخر فایل تمرین حتما به دقت مطالعه شود.
</div>

# **سوال اول:**

*بخش اول:*
---
# پاسخ خود را اینجا بنویسید:
<div style="direction: rtl; text-align: right;">
    الف:
    <br>
    NER به شناسایی و طبقه‌بندی موجودیت‌های نام‌دار در متن می‌پردازد. این موجودیت‌ها می‌توانند شامل افراد، مکان‌ها، سازمان‌ها، تاریخ‌ها، مقادیر و غیره باشند.
    <br><br>
    مراحل و اهداف NER:
    <br><br>
    شناسایی موجودیت‌ها: سیستم متن را تحلیل می‌کند و موجودیت‌های نام‌دار را شناسایی می‌کند. به عنوان مثال، در جمله "باراک اوباما رئیس‌جمهور ایالات متحده است"، "باراک اوباما" به عنوان یک شخص و "ایالات متحده" به عنوان یک مکان شناسایی می‌شود.
    <br><br>
    طبقه‌بندی موجودیت‌ها: پس از شناسایی، موجودیت‌ها به دسته‌های مختلفی مانند شخص، مکان، سازمان و غیره تقسیم‌بندی می‌شوند.
    <br><br>
    استخراج اطلاعات: با استفاده از این متد، می‌توان اطلاعات مفیدی از متن استخراج کرد که می‌تواند در تحلیل داده‌ها، خلاصه‌سازی، پاسخ به سوالات و دیگر کاربردهای پردازش زبان‌های طبیعی مفید باشد.
    <br><br>
    ------------------------------------------
    <br><br>
    ج:
    <br><br>
    مزایا:
    <br><br>
    1.	سادگی: نوشتن regex معمولاً ساده و مستقیم است و برای موارد خاصی که الگوهای واضحی دارند، بسیار کارآمد است.
    <br><br>
    2.	سرعت: regex معمولاً سریع‌تر از الگوریتم‌های پیچیده NER عمل می‌کند، به خصوص برای متون کوچک یا خاص.
    <br><br>
    3.	کنترل کامل: با استفاده از regex، توسعه‌دهنده می‌تواند کنترل دقیقی بر روی الگوهای شناسایی شده داشته باشد و به راحتی می‌تواند الگوها را تغییر دهد.
    <br><br>
    معایب:
    <br><br>
    1.	محدودیت در انعطاف‌پذیری: regex برای شناسایی الگوهای پیچیده و متنوع کارآمد نیست. به عنوان مثال، نمی‌تواند به راحتی موجودیت‌های نام‌دار با ساختارهای متفاوت را شناسایی کند.
    <br><br>
    2.	حساسیت به خطا: اگر متن کمی تغییر کند، regex ممکن است دیگر نتواند موجودیت‌ها را شناسایی کند. به عنوان مثال، تغییر در املای یک نام ممکن است باعث عدم شناسایی شود.
    <br><br>
    3.	نیاز به نگهداری: با افزایش پیچیدگی متون، نگهداری و به‌روزرسانی regex می‌تواند زمان‌بر و دشوار باشد.
    <br><br>
</div>


In [51]:
#write your code here

import re
with open("/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/ner.txt", "r") as file:
    f = file.read()
Emails = re.findall(r'[\w.+-]+@[\w-]+\.[\w.-]+', f)
Phones = re.findall(r'\++[\d+-]*+[\d+\ ]*', f)
Url = re.findall(r'https?:[^\s]+', f)

print(Emails)
print(Phones)
print(Url)

['info@vrinnovations.com', 'john.martinez@techadvisors.net', 'support@innovateads.co', 'alice.p@mailmasters.com', 'sales@mailmasters.com.', 'director@socialbuzz.org', 'support@deepad.ai', 'info@legaleagle.org', 'partners@influenceme.io', 'contact@brandvision.com', 'hello@mobilemarketersinc.com', 'paul.jenkins@cxinnovators.com', 'info@digitalfuture2025.com']
['+1-800-555-1234', '+44 20 7946 0958 ', '+1-310-555-7890', '+1-212-555-6677', '+1-718-555-9988', '+49 30 1234 5678 ', '+33 1 2345 6789', '+61 3 9876 5432', '+1-202-555-2244', '+1-415-555-3344', '+1-646-555-9988 ', '+1-888-555-1234']
['https://www.digitalfuture2025.com.', 'http://www.techadvisors.net),', 'https://www.innovateads.co)', 'https://www.innovateads.co/blog.', 'http://www.mailmasters.com)', 'https://www.socialbuzz.org),', 'http://www.deepad.ai)', 'http://www.deepad.ai/research.', 'https://www.legaleagle.org/marketing-laws),', 'http://www.influenceme.io)', 'https://www.brandvision.com)', 'http://www.mobilemarketersinc.com),

<div style="direction: rtl; text-align: right;">
    متغیر Emails رشته هایی را استخراج میکند که با یک حرف شروع می شود و بعد از آن هرچیزی به اندازه یکی یا بیشتر میتواند بیاید سپس باید یک علامت @ داشته باشد و دوباره با حروف ادامه داشته باشد، سپس یک نقطه بیاید و باز حروفی متوالی بیاید.
    <br><br>
    متغیر phones رشته هایی را استخراج میکند که با یک + شروع شود سپس تعدادی عدد بیاید 
    <br><br>
    متغیر URL رشته هایی را استخراج میکند که با http شروع شود و وجود کاراکتر s اختیاری است و بعد از آن باید یک رشته بیاید.
    <div>
    

*بخش دوم:*
---
# پاسخ خود را اینجا بنویسید:

<div style="direction: rtl; text-align: right;">
    الف:
    <br>
    توکنایزرهای مبتنی بر قواعد (Rule-based)
    <br><br>
    مزایا:
    <br><br>
    1. ساده و قابل فهم: این توکنایزرها معمولاً با استفاده از مجموعه‌ای از قواعد و الگوها عمل می‌کنند که ساده و واضح هستند.
    <br><br>
    2. سرعت بالا: به دلیل استفاده از قواعد ساده، این توکنایزرها معمولاً سریع‌تر از مدل‌های یادگیری ماشین عمل می‌کنند.
    <br><br>
    3. کنترل دقیق: توسعه‌دهنده می‌تواند به راحتی قواعد را تغییر دهد و نتایج را مطابق با نیاز خود تنظیم کند.
    <br><br>
    معایب:
    <br><br>
    1. محدودیت در انعطاف‌پذیری: این توکنایزرها به خوبی نمی‌توانند به تغییرات در زبان یا متن‌های غیرمعمول واکنش نشان دهند.
    <br><br>
    2. نیاز به نگهداری: با تغییرات زبان یا نیاز به شناسایی الگوهای جدید، باید قواعد جدیدی اضافه شود که این می‌تواند زمان‌بر باشد.
    <br><br>
    3. حساسیت به خطا: در صورتی که متن کمی تغییر کند، ممکن است توکنایزر نتواند به درستی عمل کند.
    <br><br>
    شرایط مناسب استفاده:
    <br><br>
    • در پروژه‌های ساده و با متن‌های مشخص، جایی که الگوهای واضح وجود دارد.
    <br><br>
    • زمانی که سرعت پردازش مهم باشد و نیاز به پردازش سریع متون وجود دارد.
    <br><br>
    ---------------------
    <br><br>
    توکنایزرهای مبتنی بر یادگیری ماشین (Machine Learning-based)
    <br><br>
    مزایا:
    <br>
    1. انعطاف‌پذیری بالا: این توکنایزرها می‌توانند به طور خودکار به شناسایی الگوهای پیچیده‌تر و موجودیت‌های جدید بپردازند.
    <br><br>
    2. دقت بالا: با استفاده از مدل‌های یادگیری عمیق و یادگیری ماشین، می‌توانند دقت بالایی در شناسایی توکن‌ها داشته باشند.
    <br><br>
    3. قابلیت یادگیری: این توکنایزرها می‌توانند از داده‌های آموزشی یاد بگیرند و با گذشت زمان بهبود یابند.
    <br><br>
    معایب:
    <br><br>
    1. نیاز به داده‌های آموزشی: برای آموزش مدل‌های یادگیری ماشین به داده‌های زیادی نیاز است که تهیه آن‌ها ممکن است دشوار باشد.
    <br><br>
    2. پیچیدگی محاسباتی: معمولاً به منابع محاسباتی بیشتری نسبت به توکنایزرهای مبتنی بر قواعد نیاز دارند و ممکن است زمان بیشتری برای پردازش بگیرند.
    <br><br>
    3. تنظیمات پیچیده: نیاز به تنظیمات و پارامترهای بیشتری دارد که ممکن است برای توسعه‌دهندگان تازه‌کار دشوار باشد.
    <br><br>
    شرایط مناسب استفاده:
    <br><br>
    • در پروژه‌های پیچیده و با متن‌های متنوع، جایی که الگوهای واضح وجود ندارد و متن شامل موارد نامتعارف است.
    <br><br>
    • زمانی که دقت در شناسایی توکن‌ها اهمیت دارد و داده‌های آموزشی مناسب موجود است.
    <br><br>
    انتخاب بین توکنایزرهای مبتنی بر قواعد و یادگیری ماشین بستگی به نیازهای خاص پروژه، نوع متن و منابع در دسترس دارد. در پروژه‌های ساده و با الگوهای مشخص، توکنایزرهای مبتنی بر قواعد معمولاً مناسب‌تر هستند. در حالی که در پروژه‌های پیچیده‌تر و زمانی که دقت بالا و انعطاف‌پذیری مهم باشد، توکنایزرهای مبتنی بر یادگیری ماشین بهتر عمل می‌کنند.
</div>


In [52]:
#write your code here
import re
with open("/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/ner.txt", "r") as file:
    f = file.read()

tokenized = re.findall(r'(\S+)', f)
print(tokenized)


['As', 'we', 'move', 'further', 'into', 'the', 'digital', 'age,', 'the', 'landscape', 'of', 'marketing', 'is', 'evolving', 'at', 'a', 'pace', 'faster', 'than', 'ever', 'before.', 'Businesses', 'are', 'harnessing', 'the', 'power', 'of', 'artificial', 'intelligence,', 'data', 'analytics,', 'and', 'personalized', 'advertising', 'to', 'reach', 'their', 'target', 'audiences.', 'If', 'you', 'wish', 'to', 'learn', 'more', 'about', 'the', 'latest', 'trends,', 'visit', 'https://www.digitalfuture2025.com.', 'One', 'significant', 'change', 'is', 'the', 'integration', 'of', 'augmented', 'reality', '(AR)', 'in', 'advertising', 'campaigns.', 'Brands', 'like', 'VRInnovations', '(contact:', 'info@vrinnovations.com)', 'are', 'leading', 'the', 'charge,', 'offering', 'immersive', 'experiences', 'that', 'let', 'users', '"try"', 'products', 'virtually', 'before', 'making', 'a', 'purchase.', 'This', 'level', 'of', 'engagement', 'is', 'a', 'game-changer.', 'For', 'collaborations,', 'reach', 'out', 'to', 'the

<div style="direction: rtl; text-align: right;">
    ابتدا فایل مورد نظر را میخوانیم سپس با استفاده از regex رشته های بین اسپیس ها را یافته و استخراج میکنیم.

*بخش سوم:*
---
# پاسخ خود را اینجا بنویسید:

<div style="direction: rtl; text-align: right;">
    الف:
    <br><br>
    الگوریتم‌های Levenshtein Distance و Damerau-Levenshtein Distance ابزارهایی برای اندازه‌گیری شباهت بین دو رشته (string) هستند. این الگوریتم‌ها به‌ویژه در پردازش زبان طبیعی، اصلاح خودکار متن و شناسایی الگو کاربرد دارند. در ادامه، هر یک از این الگوریتم‌ها و طرز کار آن‌ها را شرح می‌دهم.
    <br><br>
    1. Levenshtein Distance
    <br><br>
    تعریف: Levenshtein Distance فاصله ویرایش (edit distance) بین دو رشته را اندازه‌گیری می‌کند. این فاصله نشان می‌دهد که حداقل چند عملیات ویرایش (حذف، اضافه، یا جایگزینی کاراکتر) برای تبدیل یک رشته به رشته دیگر لازم است.
    <br><br>
    طرز کار:
    <br><br>
    •	عملیات‌های مجاز عبارتند از:
    <br><br>
    1.	حذف: حذف یک کاراکتر از رشته.
    <br><br>
    2.	اضافه: اضافه کردن یک کاراکتر به رشته.
    <br><br>
    3.	جایگزینی: جایگزینی یک کاراکتر با کاراکتر دیگر.
    <br><br>
    2. Damerau-Levenshtein Distance
    <br><br>
    تعریف: Damerau-Levenshtein Distance نسخه‌ای بهبود یافته از Levenshtein Distance است که علاوه بر عملیات‌های سه‌گانه فوق، عمل جابجایی (transposition) دو کاراکتر مجاور را نیز در نظر می‌گیرد.
    <br><br>
    طرز کار:
    <br><br>
    عملیات‌های مجاز عبارتند از:
    <br><br>
    1.	حذف: حذف یک کاراکتر از رشته.
    <br><br>
    2.	اضافه: اضافه کردن یک کاراکتر به رشته.
    <br><br>
    3.	جایگزینی: جایگزینی یک کاراکتر با کاراکتر دیگر.
    <br><br>
    4.	جابجایی: جابجایی دو کاراکتر مجاور.
    <br><br>
    مقایسه:
    <br><br>
    Levenshtein Distance تنها سه عملیات ویرایش را در نظر می‌گیرد، در حالی که Damerau-Levenshtein Distance چهار عملیات ویرایش را در نظر می‌گیرد.
    <br><br>
    بنابراین، Damerau-Levenshtein Distance معمولاً دقت بیشتری در شناسایی شباهت‌ها و تفاوت‌ها بین رشته‌ها ارائه می‌دهد، به‌ویژه زمانی که جابجایی کاراکترها در متن وجود دارد.
    <br><br>
    کاربردها:
    <br><br>
    تصحیح خودکار متن: شناسایی و تصحیح خطاهای املایی.
    <br><br>
    جستجوی تقریب: جستجوی رشته‌هایی که به طور تقریبی مشابه یکدیگر هستند.
    <br><br>
    تحلیل متن: ارزیابی شباهت بین متن‌ها و شناسایی تغییرات.

</div>


In [8]:
#write your code here

def levenshtein(str1, str2):
    len_str1 = len(str1) + 1 
    len_str2 = len(str2) + 1
    matrix = [[0 for _ in range (len_str2)] for _ in range (len_str1)]
    for i in range (len_str1):
        matrix [i][0] = i
    for j in range (len_str2):
        matrix [0][j] = j
    
    for i in range(1, len_str1):
        for j in range(1, len_str2):
            if str1[i-1] == str2[j-1]:
                cost = 0
            else:
                cost = 1
            matrix [i][j] = min(matrix[i-1][j] + 1,
                                matrix[i][j-1] + 1,
                                matrix[i-1][j-1] + cost)
    return matrix[len_str1-1][len_str2-1]




def damerau_levenshtein(str1, str2):
    len_str1 = len(str1) + 1 
    len_str2 = len(str2) + 1
    matrix = [[0 for _ in range (len_str2)] for _ in range (len_str1)]
    for i in range (len_str1):
        matrix [i][0] = i
    for j in range (len_str2):
        matrix [0][j] = j
    
    for i in range(1, len_str1):
        for j in range(1, len_str2):
            if str1[i-1] == str2[j-1]:
                cost = 0
            else:
                cost = 1
            matrix [i][j] = min(matrix[i-1][j] + 1,
                                matrix[i][j-1] + 1,
                                matrix[i-1][j-1] + cost)
            if i > 1 and j > 1 and str1[i - 1] == str2[j - 2] and str1[i - 2] == str2[j - 1]:
                matrix[i][j] = min(matrix[i][j], matrix[i - 2][j - 2] + 1) 
    return matrix[len_str1-1][len_str2-1]


list = [ ("kitten", "sitting"), ("saturday", "sunday"), ("book", "back"), ("algorithm",
"logarithm"), ("", "test"), ("abc", "acb") ]

for i in list:
    print('levenshtein distance between' ,i[0], 'and', i[1], 'is:', levenshtein(i[0], i[1]))


for i in list:
    print('damerau-levenshtein distance between' ,i[0], 'and', i[1], 'is:', damerau_levenshtein(i[0], i[1]))

levenshtein distance between kitten and sitting is: 3
levenshtein distance between saturday and sunday is: 3
levenshtein distance between book and back is: 2
levenshtein distance between algorithm and logarithm is: 3
levenshtein distance between  and test is: 4
levenshtein distance between abc and acb is: 2
damerau-levenshtein distance between kitten and sitting is: 3
damerau-levenshtein distance between saturday and sunday is: 3
damerau-levenshtein distance between book and back is: 2
damerau-levenshtein distance between algorithm and logarithm is: 3
damerau-levenshtein distance between  and test is: 4
damerau-levenshtein distance between abc and acb is: 1


<div style="direction: rtl; text-align: right;">
    Levenshtein
    <br><br>
    این تابع برای محاسبه فاصله Levenshtein بین دو رشته طراحی شده است. فاصله Levenshtein تعداد تغییرات (افزودن، حذف، یا جایگزینی یک کاراکتر) مورد نیاز برای تبدیل یک رشته به رشته دیگر را محاسبه می‌کند.
    <br><br>
    ابتدا طول دو رشته ورودی را به همراه ۱ به دست می‌آورد و یک ماتریس دو بعدی با ابعاد [طول رشته ۱ + ۱] در [طول رشته ۲ + ۱] ایجاد می‌کند. هر عنصر این ماتریس برای نگه‌داشتن کمترین هزینه تبدیل بین دو رشته استفاده می‌شود.
    <br><br>
    سطر و ستون اول ماتریس با اعداد از ۰ تا طول رشته‌ها پر می‌شوند، که نمایانگر هزینه تبدیل از یک رشته خالی به رشته مورد نظر است.
    <br><br>
    سپس برای هر کاراکتر دو رشته، بررسی می‌کند که آیا کاراکترهای مورد نظر برابر هستند یا خیر:
    <br><br>
    اگر برابر باشند، هزینه تبدیل ۰ است.
    <br>
    اگر برابر نباشند، هزینه ۱ به تبدیل اضافه می‌شود.
    <br><br>
    کمترین هزینه از سه عملیات (حذف، درج، یا جایگزینی) انتخاب شده و در ماتریس ذخیره می‌شود.
    <br><br>
    در پایان، آخرین مقدار ماتریس (پایین‌ترین راست) که نشان‌دهنده کمترین هزینه تبدیل کل دو رشته است، برگردانده می‌شود.
    <br><br><br>
    Damerau-Levenshtein
    <br><br>
    این تابع نیز فاصله بین دو رشته را محاسبه می‌کند، اما با یک تغییر مهم: در این متد، علاوه بر عملیات‌های درج، حذف، و جایگزینی، یک عملیات دیگر نیز مجاز است که به آن transposition می‌گویند. در این حالت، دو کاراکتر مجاور می‌توانند جابه‌جا شوند.
    <br><br>
    همانند تابع Levenshtein، اینجا نیز ابتدا ماتریس ایجاد می‌شود و سطر و ستون اول آن با اعداد پر می‌شود.
    <br><br>
    عملیات‌های مقایسه و هزینه مشابه هستند، اما اینجا یک بررسی اضافه وجود دارد: اگر کاراکترهای متوالی در دو رشته معادل هم باشند، اما ترتیب آن‌ها متفاوت باشد، می‌تواند یک هزینه کمتر (۱ واحد) برای جابه‌جایی آن‌ها در نظر بگیرد.
    <br><br>
    این ویژگی باعث می‌شود که این الگوریتم دقیق‌تر باشد، چرا که جابه‌جایی‌ها را به عنوان یک عملیات واحد در نظر می‌گیرد.
<div>

# **سوال دوم:**

*بخش اول:*
---
# پاسخ خود را اینجا بنویسید:


<div style="direction: rtl; text-align: right;">
    الف. نحوه آموزش WordPiece Tokenizer
    <br><br>
    WordPiece Tokenizer یک روش توکن‌سازی است که ابتدا در مدل‌های زبان BERT (Bidirectional Encoder Representations from Transformers) معرفی شد. این توکنایزر به‌خصوص در پردازش زبان طبیعی (NLP) برای توکن‌سازی کلمات و بهینه‌سازی فضای ذخیره‌سازی و محاسبات استفاده 
    می‌شود. نحوه آموزش این توکنایزر به‌طور خلاصه به شرح زیر است:
    <br><br>
    جمع‌آوری داده‌ها: برای آموزش یک WordPiece Tokenizer، ابتدا نیاز به مجموعه‌ای از داده‌های متنی بزرگ است. این داده‌ها معمولاً شامل متن‌های متنوع از منابع مختلف می‌شوند.
    <br><br>
    تهیه توکن‌های اولیه: با استفاده از کاراکترها و کلمات موجود در متن، توکن‌های اولیه ایجاد می‌شوند. این توکن‌ها شامل تمام کاراکترها و کلمات متنی هستند.
    <br><br>
    محاسبه فراوانی: برای هر توکن، تعداد دفعاتی که در مجموعه داده‌ها ظاهر می‌شود محاسبه می‌شود. این فراوانی‌ها به تعیین اینکه کدام توکن‌ها باید به‌عنوان توکن‌های جدید ترکیب شوند کمک می‌کند.
    <br><br>
    ایجاد ترکیبات: بر اساس فراوانی توکن‌ها، ترکیبات جدیدی از توکن‌ها ایجاد می‌شود. به‌عنوان مثال، اگر "un" و "happy" هر دو توکن‌های رایجی باشند، می‌توانند به یک توکن جدید به نام "unhappy" ترکیب شوند.
    <br><br>
    تکرار فرآیند: مراحل 3 و 4 به‌طور مکرر انجام می‌شود تا زمانی که تعداد توکن‌ها به تعداد معین یا اندازه‌ی موردنظر کاهش یابد.
    <br><br>
    ایجاد دیکشنری: در نهایت، یک دیکشنری از توکن‌ها ایجاد می‌شود که شامل تمام توکن‌های بهینه‌شده و فراوانی‌های آن‌ها است.
    <br><br>
    -----------------------------------
    <br><br>
    ب. نمونه‌ای از مدل‌های معروفی که با WordPiece Tokenizer آموزش دیده‌اند
    <br><br>
    BERT (Bidirectional Encoder Representations from Transformers): این مدل یکی از مشهورترین مدل‌های NLP است که برای بسیاری از وظایف زبان طبیعی مانند تشخیص موجودیت، تحلیل احساسات و سوال و جواب استفاده می‌شود.
    <br><br>
    ALBERT (A Lite BERT): این مدل نسخه بهینه‌شده BERT است که برای کاهش اندازه و افزایش کارایی طراحی شده است.
    <br><br>
    DistilBERT: این مدل نسخه فشرده شده BERT است که برای حفظ دقت و کارایی بهینه‌سازی شده است.
    <br><br>
    -------------------------------------
    <br><br>
    ج. مقایسه WordPiece و BPE
    <br><br>
    WordPiece: این روش به‌خصوص برای حل مشکلات توکن‌سازی در زبان‌هایی که واژه‌ها دارای ترکیب‌های مختلفی هستند، مناسب است. WordPiece از یک روش مبتنی بر احتمال استفاده می‌کند و توکن‌ها را با توجه به فراوانی آن‌ها در متن ایجاد می‌کند. این توکنایزر می‌تواند به‌طور مداوم توکن‌های جدیدی بسازد و به راحتی با زبان‌های مختلف سازگار است.
    <br><br>
    BPE (Byte Pair Encoding): این روش بر اساس ترکیب جفت‌های کاراکتر متوالی و تبدیل آن‌ها به یک توکن جدید کار می‌کند. BPE نیز به‌خوبی با زبان‌های مختلف سازگار است و می‌تواند واژه‌ها را به زیرواژه‌ها تقسیم کند، اما معمولاً در کاربردهای محدودتری نسبت به WordPiece استفاده می‌شود.
</div>


In [14]:
#write your code here
from tokenizers import Tokenizer, models, trainers, pre_tokenizers
tokenizer = Tokenizer(models.WordPiece(unk_token="[UNK]"))
tokenizer.pre_tokenizer = pre_tokenizers.Whitespace()
trainer = trainers.WordPieceTrainer(
    vocab_size=30000, 
    min_frequency=2, 
    special_tokens=["[UNK]"]
)

with open("/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/ferdowsi.txt", "r", encoding='utf-8') as file:
    text = file.read()

tokenizer.train_from_iterator([text], trainer)

tokenizer.save("wordpiece_tokenizer.json")

tokenizer = Tokenizer.from_file("wordpiece_tokenizer.json")

output = tokenizer.encode("سلام این یک جمله جهت تست توکنایزر است")
print("Tokens:", output.tokens)




Tokens: ['سلام', 'این', 'یک', 'جمله', 'جه', '##ت', 'تست', 'تو', '##کن', '##ای', '##زر', 'است']


<div style="direction: rtl; text-align: right;">
    این کد نخست کتابخانه های مورد نیاز را ایمپورت میکند.
    <br><br>
    سپس  یک مدل WordPiece ایجاد می‌شود که توکن [UNK] را برای کلماتی که در واژگان حضور ندارند، به‌کار می‌برد. این مدل برای شکستن کلمات به بخش‌های کوچک‌تر استفاده می‌شود که در فرآیند توکنیزیشن کاربرد دارد.
    <br><br>
    در مرحله بعد از یک پیش توکنایزر مبتنی بر فاصله  استفاده می‌شود تا کلمات موجود در متن بر اساس فاصله‌ها از هم جدا شوند. این کار باعث می‌شود که هر کلمه به صورت مجزا به عنوان یک توکن مورد بررسی قرار گیرد.
    <br><br>
    سپس یک ترینر تنظیم میشود که مسئولیت یادگیری مدل را بر عهده دارد. این ترینر پارامترهایی مانند اندازه واژگان (۳۰,۰۰۰ کلمه)، حداقل فراوانی کلمات (۲ بار تکرار در متن) و توکن‌های ویژه (مثل [UNK]) را برای آموزش مدل تنظیم می‌کند. 
    <br><br>
    در مرحله بعد فایل ferdowsi.txt خوانده میشود و مدل با استفاده از متنی که خوانده شده آموزش می بیند و بعد از آموزش مدل توکنایزر در فایلی با نام wordpiece_tokenizer.json ذخیر میشود.
    <br><br>
    سپس مدل توکنیازر ذخیره شده بازیابی میشود و با یک جمله تست میشود.
    <br><br>
    جمله مورد نظر "سلام،این یک جمله جهت تست توکنایزر است" میباشد که به توکن های زیر تقسیم میشود:
    <br>
    Tokens: ['سلام', 'این', 'یک', 'جمله', 'جه', '##ت', 'تست', 'تو', '##کن', '##ای', '##زر', 'است']
<div>


*بخش سوم:*
---
# پاسخ خود را اینجا بنویسید:


<div style="direction: rtl; text-align: right;">
    ج
    <br><br>
    1. فضای ذخیره‌سازی بزرگ
    <br><br>
    با افزایش مقدار n n، تعداد ترکیب‌های ممکن از توکن‌ها به‌طرز چشمگیری افزایش می‌یابد. به‌عنوان مثال، در یکمجموعه داده بزرگ، ممکن است نیاز به ذخیره‌سازی میلیون‌ها یا حتی میلیاردها n-gram وجود داشته باشد. این امر به‌ویژه در مدل‌های بزرگ و پیچیده، به منابع حافظه زیادی نیاز دارد.
    <br><br>
    2. کمبود داده (Sparsity)
    <br><br>
    در n-gram های بزرگ، بسیاری از ترکیب‌ها ممکن است در داده‌های آموزشی وجود نداشته باشند. این کمبود داده می‌تواند منجر به مشکل "کمبود داده" (data sparsity) شود، که به این معناست که n-gram های خاصی ممکن است به‌اندازه کافی اطلاعات نداشته باشند تا بتوانند پیش‌بینی‌های دقیقی انجام دهند.
    <br><br>
    3. مدل‌سازی وابستگی‌های بلندمدت
    <br><br>
    n-gram ها معمولاً به روابط محلی بین توکن‌ها توجه دارند و در مدل‌سازی وابستگی‌های بلندمدت (long-range dependencies) ناتوان هستند. به همین دلیل، آن‌ها نمی‌توانند اطلاعات مربوط به توالی‌های طولانی‌تر را به‌خوبی پردازش کنند، که این مسأله می‌تواند در مواردی که متن به شدت وابسته است، منجر به نتایج نادرست شود.
    <br><br>
    4. عدم قطعیت و تنوع
    <br><br>
    n-gram ها قادر به مدل‌سازی عدم قطعیت در پیش‌بینی‌ها نیستند. به‌عنوان مثال، ممکن است یک ترکیب n-gram در یک متن مشخص وجود نداشته باشد، اما مدل بر اساس توکن‌های نزدیک‌تر، پیش‌بینی‌های نادرستی انجام دهد. این به این معناست که ن-گرام ها نمی‌توانند تنوع و تطبیق‌های مختلف در زبان را به‌خوبی در نظر بگیرند.
    <br><br>
    5. توزیع‌های غیر یکنواخت
    <br><br>
    در بسیاری از زبان‌ها، برخی از توکن‌ها به‌طور طبیعی بیشتر از دیگر توکن‌ها ظاهر می‌شوند. این توزیع‌های غیر یکنواخت می‌تواند منجر به بایاس (bias) در مدل شود و باعث شود که برخی از توکن‌ها یا n-gram ها به‌طور غیرمنصفانه‌ای بر نتایج تأثیر بگذارند.
    <br><br>
    6. هزینه محاسباتی
    <br><br>
    افزایش مقدار 𝑛n منجر به افزایش هزینه‌های محاسباتی می‌شود. در هنگام پیش‌بینی، مدل باید تمام n-gram های ممکن را بررسی کند که این فرآیند می‌تواند زمان‌بر باشد و منابع محاسباتی زیادی را مصرف کند.
    <br><br>
    7. محدودیت در پیش‌بینی
    <br><br>
    n-gram ها محدود به پیش‌بینی توکن‌ها بر اساس تعداد ثابتی از توکن‌های قبلی هستند. به همین دلیل، این روش‌ها در مسائلی مانند تولید متن یا تولید زبان طبیعی با تنوع و پیچیدگی بالا با چالش‌هایی روبه‌رو می‌شوند.
</div>


In [3]:
#write your code here
#write your code here
import collections
import random
from tokenizers import Tokenizer

class NgramModel:
    def __init__(self, n, tokenizer_path):
        self.n = n
        self.ngrams = collections.defaultdict(collections.Counter)
        self.vocab = set()
        self.tokenizer = Tokenizer.from_file(tokenizer_path)

    def tokenize(self, text):
        return self.tokenizer.encode(text).tokens

    def build_ngrams(self, tokens):
        for i in range(len(tokens) - self.n + 1):
            gram = tuple(tokens[i:i + self.n - 1])
            next_word = tokens[i + self.n - 1]
            self.ngrams[gram][next_word] += 1
            self.vocab.add(next_word)

    def train(self, text):
        tokens = self.tokenize(text)
        self.build_ngrams(tokens)
        print(f"Trained on {len(tokens)} tokens, built {sum(len(v) for v in self.ngrams.values())} n-grams.")

    def predict(self, context):
        context = tuple(context)
        possible_next_words = self.ngrams.get(context, {})

        if possible_next_words:
            vocab_size = len(self.vocab)
            total_count = sum(possible_next_words.values()) + vocab_size  # Add-1 smoothing
            # Calculate smoothed probabilities
            words, probs = zip(*[(word, (count + 1) / total_count) for word, count in possible_next_words.items()])
            
            # Selecting next word based on probabilities
            return random.choices(words, weights=probs)[0]
        else:
            # Fallback to random choice if no context found
            return random.choice(list(self.vocab))  # Ensure list is not redefined

    def generate_text(self, seed_text, length):
        generated = seed_text[:]
        context = seed_text[-(self.n - 1):]
        for _ in range(length - len(seed_text)):
            next_word = self.predict(context)
            generated.append(next_word)
            context = generated[-(self.n - 1):]
        return " ".join(generated)

tokenizer_path = "/Users/erfanshahabi/Desktop/uni/NLP/CA1/wordpiece_tokenizer.json"

models = {
    "2-gram": NgramModel(2, tokenizer_path),
    "4-gram": NgramModel(4, tokenizer_path),
    "8-gram": NgramModel(8, tokenizer_path)
}

with open("/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/ferdowsi.txt", "r", encoding='utf-8') as file:
    text = file.read()

for name, model in models.items():
    print(f"Training {name}...")
    model.train(text)

seed_text = ["هنر", "همانا"]  # بذر برای شروع متن
length = 200  # طول متن تولیدی

for name, model in models.items():
    print(f"\nGenerating text with {name} model:")
    generated_text = model.generate_text(seed_text, length)
    print(generated_text)



Training 2-gram...
Trained on 671364 tokens, built 186869 n-grams.
Training 4-gram...
Trained on 671364 tokens, built 579111 n-grams.
Training 8-gram...
Trained on 671364 tokens, built 666531 n-grams.

Generating text with 2-gram model:
هنر همانا ز داد و سرزنش نشنوند | اگر خواهد گزید | ز یزدان مر | مراین را بپیچید سرت گر نامور مای و هیچ دم | از نیاکان برترمنش | نگه کرد یاد | ایا باد آفریدون وز درد | گزیده ز هنگام جوی تخت بنشاختشان | ببند | کنون شاه و ز خاور خدای بزرگ | سپرهای چینی نه گوپال باید و نام دارم ##ت مالید و زیبا شدست | همه شاه سترگ | به زرین کمر | بذرگشسب و سود آید به گرد اندرش | بزرگان بسود ##ه شد روشنایی ببرد آن خوشن ##وار بود و پرنیانی درفش | بدو گفت فرمان همی | که گرگ | زبانها پر درد او داشتی | ز زابلستان خرامید یک‌یک به بربر بفرمان ببستند بر خویشتن را به دام بود شاد شد سبک | که هست از ##گوش | به زه زهازه گرفت این کشور و با چنین تا چشمهٔ عنبر گرفت | به منقار مرگ ما | ازین برف ##کند گاو | بپرسیدن از درد | کسی کو بجایی توان و خوبان آراسته | ازیشان گزیده سواری و بوی بر خاک و

<div style="direction: rtl; text-align: right;">
    این کد یک مدل N-گرم (N-gram Model) را برای تولید متن بر اساس توکن‌های ورودی پیاده‌سازی می‌کند.
    <br><br>
    کد از collections برای مدیریت n-گرم‌ها، random برای انتخاب تصادفی کلمات، و tokenizers برای توکنایز کردن متن استفاده می‌کند.
    <br><br>
    کلاس NgramModel شامل متدها و ویژگی‌های زیر است:
    <br><br>
    متغیرها:
    <br><br>
    n: تعداد n-گرم‌ها.
    <br>
    ngrams: دیکشنری برای نگهداری n-گرم‌ها و تعداد وقوع آن‌ها.
    <br>
    vocab: مجموعه‌ای از کلمات موجود در متن.
    <br>
    tokenizer: مدل توکنایزر بارگذاری شده.
    <br><br>
    متدها:
    <br><br>
    __init__: برای مقداردهی اولیه متغیرها و بارگذاری توکنایزر.
    <br>
    tokenize: ورودی متنی را توکنایز کرده و لیست توکن‌ها را برمی‌گرداند.
    <br>
    build_ngrams: n-گرم‌ها را با استفاده از توکن‌ها ایجاد می‌کند.
    <br>
    train: مدل را با استفاده از متن ورودی آموزش می‌دهد و تعداد توکن‌ها و n-گرم‌های ایجاد 
    شده را چاپ می‌کند.
    <br>
    predict: بر اساس زمینه (context) یک کلمه پیش‌بینی می‌کند. از تکنیک "Add-1 smoothing" برای محاسبه احتمال کلمات استفاده می‌شود.
    <br>
    generate_text: متنی با طول مشخص را بر اساس یک متن بذر (seed text) تولید می‌کند.
    <br><br>
    در مرحله بعد سه مدل n-گرم (2-گرم، 4-گرم و 8-گرم) با استفاده از مسیر توکنایزر مشخص شده ایجاد شده‌اند.
    <br><br>
    متن از فایل ferdowsi.txt خوانده می‌شود و هر سه مدل بر روی آن آموزش می‌بینند. پیام‌هایی حاوی تعداد توکن‌های آموزش دیده و تعداد n-گرم‌های ساخته شده در حین آموزش چاپ می‌شود.
    <br><br>
    متن تولیدی با استفاده از هر سه مدل با توجه به متن بذر "هنر", "همانا" و طول مشخص ۲۰۰ کلمه تولید می‌شود. نتایج برای هر مدل چاپ می‌شود.
    <br><br>
    این قسمت از گزارش و صحت سنجی کد با استفاده از chatgpt نگارش شده است.
    <br>
    پرامپت: https://chatgpt.com/share/67126371-52cc-800d-8d41-d3e23235c68f
<div>



*بخش چهارم:*
---
# پاسخ خود را اینجا بنویسید:

<div style="direction: rtl; text-align: right;">
    الف
    <br><br>
    معیار perplexity  نشان‌دهنده این است که مدل چقدر در پیش‌بینی توزیع احتمالات کلمات یک متن خوب عمل می‌کند. به طور دقیق‌تر، perplexity یک اندازه‌گیری از میزان عدم قطعیت یا مدل است نسبت به داده‌هایی که مشاهده می‌کند.
    <br><br>
    Perplexity کم نشان می‌دهد که مدل احتمال بالاتری به توالی واقعی کلمات داده است، یعنی مدل به خوبی توانسته است کلمات را پیش‌بینی کند. این به معنای کیفیت بالای مدل در درک ساختار زبانی داده‌ها است.
    </div>

In [4]:
#write your code here
#write your code here
import collections
import random
import math
from tokenizers import Tokenizer

class NgramModel:
    def __init__(self, n, tokenizer_path):
        self.n = n
        self.ngrams = collections.defaultdict(collections.Counter)
        self.vocab = set()
        self.tokenizer = Tokenizer.from_file(tokenizer_path)

    def tokenize(self, text):
        return self.tokenizer.encode(text).tokens

    def build_ngrams(self, tokens):
        for i in range(len(tokens) - self.n + 1):
            gram = tuple(tokens[i:i + self.n - 1])
            next_word = tokens[i + self.n - 1]
            self.ngrams[gram][next_word] += 1
            self.vocab.add(next_word)

    def train(self, text):
        tokens = self.tokenize(text)
        self.build_ngrams(tokens)
        print(f"Trained on {len(tokens)} tokens, built {sum(len(v) for v in self.ngrams.values())} n-grams.")

    def predict(self, context):
        context = tuple(context)
        possible_next_words = self.ngrams.get(context, {})

        if possible_next_words:
            vocab_size = len(self.vocab)
            total_count = sum(possible_next_words.values()) + vocab_size  # Add-1 smoothing
            # Calculate smoothed probabilities
            words, probs = zip(*[(word, (count + 1) / total_count) for word, count in possible_next_words.items()])
            
            # Selecting next word based on probabilities
            return random.choices(words, weights=probs)[0]
        else:
            # Fallback to random choice if no context found
            return random.choice(list(self.vocab))  # Ensure list is not redefined

    def generate_text(self, seed_text, length):
        generated = seed_text[:]
        context = seed_text[-(self.n - 1):]
        for _ in range(length - len(seed_text)):
            next_word = self.predict(context)
            generated.append(next_word)
            context = generated[-(self.n - 1):]
        return " ".join(generated)
    def perplexity(self, text):
        tokens = self.tokenize(text)
        if len(tokens) < self.n:
            return float('inf')  # اگر تعداد توکن‌ها کمتر از n باشد، perplexity بی‌نهایت می‌شود
        log_prob_sum = 0
        N = len(tokens) - self.n + 1

        for i in range(N):
            context = tuple(tokens[i:i + self.n - 1])
            next_word = tokens[i + self.n - 1]
            if next_word in self.ngrams[context]:
                count = self.ngrams[context][next_word]
                prob = (count + 1) / (sum(self.ngrams[context].values()) + len(self.vocab))  # Add-1 smoothing
                log_prob_sum += math.log(prob)
            else:
                log_prob_sum += math.log(1 / (sum(self.ngrams[context].values()) + len(self.vocab)))  # اگر کلمه‌ای در n-gram وجود نداشته باشد

        return math.exp(-log_prob_sum / N)  # محاسبه perplexity



tokenizer_path = "/Users/erfanshahabi/Desktop/uni/NLP/CA1/wordpiece_tokenizer.json"
model = NgramModel(4, tokenizer_path) 

with open("/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/ferdowsi.txt", "r", encoding='utf-8') as file:
    text = file.read()

model.train(text)

hafez_perplexity = []  
with open("/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/hafez.txt", "r", encoding='utf-8') as file:
    for line in file:
        perplexity = model.perplexity(line.strip())
        hafez_perplexity.append(perplexity)

mean_hafez_perplexity = sum(hafez_perplexity) / len(hafez_perplexity)
print(f"Mean Perplexity for Hafez: {mean_hafez_perplexity}")

new_poems_perplexity = []
with open("/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/modern_poet.txt", "r", encoding='utf-8') as file:
    for line in file:
        perplexity = model.perplexity(line.strip())
        new_poems_perplexity.append(perplexity)

mean_new_poems_perplexity = sum(new_poems_perplexity) / len(new_poems_perplexity)
print(f"Mean Perplexity for Modern Poems: {mean_new_poems_perplexity}")


print(f"Comparison: Hafez vs Modern Poems")
print(f"Hafez Perplexity: {mean_hafez_perplexity}")
print(f"Modern Poems Perplexity: {mean_new_poems_perplexity}")


Trained on 671364 tokens, built 579111 n-grams.
Mean Perplexity for Hafez: inf
Mean Perplexity for Modern Poems: 12917.375226981987
Comparison: Hafez vs Modern Poems
Hafez Perplexity: inf
Modern Poems Perplexity: 12917.375226981987


<div style="direction: rtl; text-align: right;">
    در این کد مانند کد قبلی کلاس ngrammodel تعریف میشود و در این کلاس متد perplexity را تعریف میکنیم.
    <br><br>
    ابن متد به منظور محاسبه پیچیدگی یا عدم قطعیت مدل زبانی در پیشبینی توکن ها استفاده میشود.
    <br><br>
    این متد یک متن ورودی شامل توکن ها را دریافت میکند. سپس متن ورودی را با استفاده از توکنایزر تعریف شده در کلاس توکنایز میکند. ابتدا بررسی میشود که آیا تعداد توکن ها کمتر از n (که در اینجا تعداد توکن ها در مدل n گرم است) هست یا نه. اگر تعداد توکن ها کمتر باشد، مقدار perplexity بی نهایت یا infinity درنظر گرفته میشود چون مدل نمیتواند پیش بینی دقیقی داشته باشد.
    <br><br>
    متد به طور تکراری برای هر توکن در متن (به جز آخرین n-1 توکن) به محاسبه احتمال توکن بعدی می‌پردازد:
    <br><br>
    برای هر توکن، context یا محیط آن (n-1 توکن قبلی) تعیین می‌شود.
    <br><br>
    تعداد وقوع توکن بعدی (next_word) در n-grams مدل بررسی می‌شود.
    <br><br>
    احتمال این توکن با استفاده از فرمول Add-1 smoothing محاسبه می‌شود تا از صفر شدن احتمال جلوگیری شود. این کار باعث می‌شود که اگر توکن مورد نظر هرگز در n-grams دیده نشده باشد، 
    احتمال آن صفر نشود و به جای آن مقدار کوچکی به آن نسبت داده شود.
    <br><br>
    لگاریتم احتمال‌ها برای توکن‌های پیش‌بینی‌شده جمع‌آوری می‌شود. این کار به دلیل اینکه مقادیر احتمالی معمولاً کوچک هستند و جمع آن‌ها می‌تواند منجر به overflow شود، انجام می‌گیرد.
    <br><br>
    در این کد معیار perplexity برای اشعار حافظ برابر بی نهایت و برای اشعار مدرن برابر12917 شده است که معیار خوبی نیست چرا که perplexity هر چه به یک نزدیک تر باشد بهتر است.
<div>


# **سوال سوم:**

*بخش اول:*
---
# پاسخ خود را اینجا بنویسید:
<div style="direction: rtl; text-align: right;">
    الف
    <br><br>
    می‌توان از ngram به عنوان یک ویژگی برای ساخت مدل طبقه‌بندی استفاده کرد. در این حالت، توالی‌های n-تایی از کلمات یا کاراکترها (ngram) به عنوان ویژگی‌های ورودی به مدل طبقه‌بند داده می‌شوند. فرآیند به این صورت است که ابتدا ngramهای داده‌های متنی استخراج می‌شوند و سپس این ngramها به یک نمایش عددی (مانند بردار) تبدیل می‌شوند که قابل استفاده در الگوریتم‌های یادگیری ماشین باشد.
    <br><br>
    مراحل استفاده از ngram به عنوان classifier:
    <br><br>
    استخراج ngram: از داده‌های متنی خود ngramهایی با اندازه‌های مختلف (مثلاً 2gram یا 3gram) استخراج می‌کنید.
    <br><br>
    بردارسازی ngramها: ngramهای استخراج‌شده را می‌توان با روش‌هایی مانند Bag of ngrams یا TF-IDF به بردارهای عددی تبدیل کرد.
    <br><br>
    استفاده از الگوریتم طبقه‌بندی: بردارهای ngram را می‌توانید به الگوریتم‌های مختلف طبقه‌بندی مانند Logistic Regression، SVM، یا Random Forest بدهید تا مدل طبقه‌بندی ساخته شود.
    <br><br>
    ارزیابی مدل: پس از آموزش مدل، می‌توان معیارهایی مانند دقت، precision، recall و F1-score را برای ارزیابی عملکرد مدل محاسبه می شود.
    <br><br>
    ------------------------
    <br><br>
    ب
    <br><br>
    افزایش داده در مدل‌های n-gram برای طبقه‌بندی در صورتی که داده‌ها کم باشند، می‌تواند مشکلات زیر را به همراه داشته باشد:
    <br><br>
    افزایش فضای جستجو و تعداد پارامترها: با افزایش اندازه n-gram، تعداد ترکیب‌های ممکن از n کلمه بیشتر می‌شود. اگر داده‌های کافی برای پوشش دادن تمام این ترکیب‌ها وجود نداشته باشد، مدل به دلیل کمبود نمونه‌های کافی، نمی‌تواند ترکیب‌های مناسب را یاد بگیرد و اغلب به بایاس بیش از حد دچار می‌شود.
    <br><br>
    کمبود پوشش برای کلمات نادر یا ترکیبات نادر: در یک مجموعه داده کم، بسیاری از ترکیبات نادر ممکن است اصلاً در داده آموزشی ظاهر نشوند یا بسیار کم رخ دهند. این مشکل باعث می‌شود که مدل در مواجهه با ترکیبات جدید یا نادر عملکرد ضعیفی داشته باشد، چرا که نمی‌تواند الگوهای معنایی یا نحوی مناسب را از داده‌ها استخراج کند.
    <br><br>
    Overfitting: با افزایش تعداد n-gramها و کم بودن داده‌ها، مدل به شدت به داده‌های آموزش وابسته می‌شود و الگوهای خاصی که ممکن است در داده‌های جدید یا آزمایشی وجود نداشته باشند را بیش از حد یاد می‌گیرد. این منجر به افزایش خطای مدل در داده‌های جدید می‌شود.
    <br><br>
    افزایش پیچیدگی محاسباتی: تعداد n-gramهای ممکن به طور نمایی افزایش می‌یابد که باعث افزایش هزینه‌های محاسباتی و نیاز به حافظه می‌شود. اگر داده کافی وجود نداشته باشد، تلاش برای استفاده از n-gramهای بزرگتر فقط هزینه محاسباتی را افزایش می‌دهد بدون این که بهبود قابل توجهی در عملکرد مدل حاصل شود.
    <div>

In [5]:
#write your code here
import pandas as pd
from tokenizers import Tokenizer, models, trainers, pre_tokenizers

df = pd.read_csv("/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/digikala.csv")
texts = df['Text'].fillna("").astype(str).tolist()
suggestions = df['Suggestion'].fillna("").astype(str).tolist()
combined_texts = [text + " " + suggestion for text, suggestion in zip(texts, suggestions)]
tokenizer = Tokenizer(models.BPE())
tokenizer.pre_tokenizer = pre_tokenizers.Whitespace()
trainer = trainers.BpeTrainer(vocab_size=30000, min_frequency=2)
tokenizer.train_from_iterator(combined_texts, trainer)
tokenizer.save("/Users/erfanshahabi/Desktop/uni/NLP/CA1/bpe_tokenizer.json")
tokenizer = Tokenizer.from_file("/Users/erfanshahabi/Desktop/uni/NLP/CA1/bpe_tokenizer.json")


for text in combined_texts:
    tokens = tokenizer.encode(text)
    print("Tokens:", tokens.tokens)     
   





Tokens: ['این', 'اولین', 'تجربه', 'من', 'برای', 'خرید', 'ایفون', 'هست', 'امروز', 'ب', 'دستم', 'رسیده', 'به', 'اسم', 'طلایی', 'هست', 'ولی', 'در', 'اصل', 'رز', 'گل', 'ده', 'خیلی', 'هم', 'خوش', 'رنگ', 'هست', '1']
Tokens: ['خرید', 'این', 'محصول', 'رو', 'توصیه', 'میکنم', '1']
Tokens: ['1', 'ساله', 'این', 'گوشی', 'رو', 'دارم', 'هیچ', 'نقطه', 'ضعفی', 'ازش', 'ندیدم', '1']
Tokens: ['سلام', 'خدمت', 'دوستان', 'این', 'گوشی', 'از', 'همه', 'نظر', 'عالی', 'کیفیت', 'ساخت', 'سرعت', 'اجرا', 'کیفیت', 'صدا', 'و', 'تصویر', 'از', 'نظر', 'حافظه', 'هم', 'فقط', '5', 'گیگ', 'در', 'دسترس', 'شما', 'قرار', 'میگیره', '1']
Tokens: ['سلام', 'دوستانی', 'که', 'نگران', 'شکستن', 'صفحه', 'نمایش', 'هستند', 'این', 'رو', 'بدون', 'ند', 'که', 'صفحه', 'با', 'گوری', 'لا', 'گلس', '5', 'محافظت', 'میشه', 'و', 'با', 'خدمات', 'ویژه', 'سامسونگ', 'اگر', 'تحت', 'هر', 'شرایط', 'صفحه', 'گوشی', 'دچار', 'ایراد', 'بشه', 'با', '200', 'هزار', 'تومان', 'صفحه', 'رو', 'براتون', 'تعویض', 'میکنند', 'و', 'از', 'لحاظ', 'امکانات', 'هم', 'گوشی', 'عا

<div style="direction: rtl; text-align: right;">
    این کد یک توکنایزر bpe را برای پردازش متون استخراج شده از داده های نظرات دیجی کالا تولید میکند. در این کد از کتاب خانه pandas برای مدیریت داده های جدولی و از کتابخانه tokenizers برای ایجاد توکنایزر استفاده می شود.
    <br><br>
    بعد از ایمپورت کرد کتابخانه ها با استفاده از تابع read_csv، داده‌های مربوط به نظرات کاربران از فایل CSV بارگذاری می‌شود. این داده‌ها شامل دو ستون اصلی هستند: text و suggestion.
    <br><br>
    در مرحله بعد ستون‌های Text و Suggestion از DataFrame استخراج و مقادیر خالی آن‌ها با رشته‌های خالی جایگزین می‌شود. سپس، متن‌ها و پیشنهادات به یکدیگر ترکیب می‌شوند تا یک لیست جدید از جملات ترکیبی ایجاد شود. این کار به توکنایزر کمک می‌کند تا با متونی که شامل نظرات و پیشنهادات هستند، بهتر آموزش ببیند.
    <br><br>
    پس از آماده‌سازی داده‌ها، یک توکنایزر BPE جدید ایجاد می‌شود. توکنایزر به گونه‌ای تنظیم می‌شود که از فضای خالی به عنوان نشانه‌گذار برای توکن‌ها استفاده کند. این مرحله به توکنایزر کمک می‌کند تا متون را به توکن‌های مجزا تقسیم کند.
    <br><br>
    سپس توکنایزر با استفاده از داده‌های ترکیبی آموزش داده می‌شود. در این مرحله، پارامترهایی مانند اندازه دیکشنری (vocab size) و حداقل تعداد تکرار برای هر توکن مشخص می‌شود. پس از اتمام آموزش، توکنایزر BPE به صورت یک فایل JSON ذخیره می‌شود.
    در نهایت، توکنایزر بر روی متون ترکیبی آزمایش می‌شود و توکن‌های تولیدشده برای هر متن به نمایش گذاشته می‌شوند.
<dev>

*بخش سوم:*
---
# پاسخ خود را اینجا بنویسید:
<div style="direction: rtl; text-align: right;">
    د
    2-gram - Accuracy: 0.7274119448698315, Precision: 0.8018301562397242, Recall: 0.7274119448698315
    3-gram - Accuracy: 0.7289433384379785, Precision: 0.8026407717659014, Recall: 0.7289433384379785
    مدل3-gram دقت بهتری دارد.
    <br><br>
    مقادیر به دست آمده نشان دهنده این است که:
    <br><br>
    برای مدل 2-gram:
    <br><br>
    Accurscy برابر 0.727 است که هرچه به ۱ نزدیک تر باشد نشان از دقت بالاتر دارد و این مورد خوب است.
    <br><br>
    Precision برابر 0.801 است که نشان میدهد ۸۰ درصد مقادیر مثبتی که پیش بینی شده واقعا مثبت بودند و هر جه این معیار به ۱ نزدیک تر باشد بهتر است.
    <br><br>
    Recall برابر 0.727 است که به معنای این است که تمام نمونه‌های مثبت واقعی به درستی شناسایی شده‌اند. هر چه این مقدار به ۱ نزدیک تر باشد بهتر است.
    <br><br>
    ------------------------------
    <br><br>
    برای مدل 3-gram:
    <br><br>
    Accurscy برابر 0.728 است که هرچه به ۱ نزدیک تر باشد نشان از دقت بالاتر دارد و این مورد خوب است.
    <br><br>
    Precision برابر 0.802 است که نشان میدهد ۸۰ درصد مقادیر مثبتی که پیش بینی شده واقعا مثبت بودند و هر جه این معیار به ۱ نزدیک تر باشد بهتر است.
    <br><br>
    Recall برابر 0.728 است که به معنای این است که تمام نمونه‌های مثبت واقعی به درستی شناسایی شده‌اند. هر چه این مقدار به ۱ نزدیک تر باشد بهتر است.
    <br><br>
    در کل این مقادیر برای مدل 3-gram بیشتر است و نشان دهنده عملکرد بهتر مدل 3-gram دارد.
    <br><br>
    --------------------------------------------------------------
    <br><br>
    ه
    <br><br>
    استفاده از طبقه‌بندهای n-gram به جای طبقه‌بندهای با پارامترهای زیاد (مانند شبکه‌های عصبی عمیق یا سایر الگوریتم‌های پیچیده) در شرایط خاصی می‌تواند مفید باشد:
    <br><br>
    داده‌های محدود: وقتی که اندازه داده‌ها محدود است، استفاده از مدل‌های ساده‌تری مانند n-gram که پارامترهای کمتری دارند، می‌تواند به جلوگیری از مشکل overfitting کمک کند. مدل‌های پیچیده‌تر ممکن است به داده‌های آموزشی بیش از حد حساس شوند و در نتیجه در داده‌های جدید عملکرد ضعیفی داشته باشند.
    <br><br>
    تفسیرپذیری: n-gram به دلیل سادگی‌اش، قابلیت تفسیر بالایی دارد. می‌توان به راحتی فهمید که کدام ویژگی‌ها (یا n-gramها) بیشترین تأثیر را در پیش‌بینی‌ها دارند. این ویژگی برای کاربردهایی که نیاز به توضیح و تفسیر دارند، بسیار مهم است.
    <br><br>
    زمان محاسباتی: n-gram معمولاً سریع‌تر و کارآمدتر از مدل‌های پیچیده است. در مواقعی که زمان محاسباتی مهم باشد، استفاده از n-gram می‌تواند مزیت داشته باشد.
    <br><br>
    مسائل خطی: اگر داده‌ها به طور ذاتی خطی یا تقریباً خطی باشند، مدل‌های n-gram می‌توانند به خوبی عمل کنند و نیاز به مدل‌های پیچیده‌تر نخواهد بود.
    <br><br>
    زبان‌های با ساختار ساده: در زبان‌هایی که ساختار زبانی پیچیده‌ای ندارند یا در مسائلی مانند تحلیل متن، n-gramها می‌توانند به خوبی نشان‌دهنده اطلاعات موجود در داده‌ها باشند.
    <br><br>
    مقایسه‌ی سریع: n-gram به عنوان یک پایه برای مقایسه مدل‌های دیگر مفید است. می‌توان از n-gram به عنوان یک روش پایه استفاده کرد و سپس عملکرد مدل‌های پیچیده‌تر را نسبت به آن ارزیابی کرد.
    <br><br>
    استفاده از n-gram به ویژه در تحلیل متون و پردازش زبان طبیعی، در مواقعی که به سادگی، سرعت، و تفسیرپذیری نیاز است، می‌تواند بسیار مؤثر باشد.
    <div>

In [6]:
#write your code here
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score
from tokenizers import Tokenizer
from sklearn.feature_extraction.text import TfidfVectorizer

class NgramClassifier:
    def __init__(self, n, tokenizer_path):
        self.tokenizer = Tokenizer.from_file(tokenizer_path)  
        self.vectorizer = TfidfVectorizer(ngram_range=(n, n), max_df=0.95, min_df=2, stop_words='english')
        self.model = MultinomialNB()

    def tokenize_text(self, text):
        return " ".join(self.tokenizer.encode(text).tokens) 

    def train(self, X_train, y_train):
        X_train_tokenized = X_train.apply(self.tokenize_text)
        X_train_counts = self.vectorizer.fit_transform(X_train_tokenized)
        self.model.fit(X_train_counts, y_train)

    def evaluate(self, X_test, y_test):
        X_test_tokenized = X_test.apply(self.tokenize_text)
        X_test_counts = self.vectorizer.transform(X_test_tokenized)
        y_pred = self.model.predict(X_test_counts)
        acc = accuracy_score(y_test, y_pred)
        precision = precision_score(y_test, y_pred, average='weighted', zero_division=1)
        recall = recall_score(y_test, y_pred, average='weighted')
        return acc, precision, recall

data = pd.read_csv('/Users/erfanshahabi/Desktop/uni/NLP/CA1/NLP_CA1/data/digikala.csv')

data.dropna(subset=['Text', 'Suggestion'], inplace=True)
train_data, test_data = train_test_split(data, test_size=0.2, random_state=42)
X_train, y_train = train_data['Text'], train_data['Suggestion']
X_test, y_test = test_data['Text'], test_data['Suggestion']

ngram_2 = NgramClassifier(2, "/Users/erfanshahabi/Desktop/uni/NLP/CA1/bpe_tokenizer.json")
ngram_2.train(X_train, y_train)
acc_2, precision_2, recall_2 = ngram_2.evaluate(X_test, y_test)

ngram_3 = NgramClassifier(3, "/Users/erfanshahabi/Desktop/uni/NLP/CA1/bpe_tokenizer.json")
ngram_3.train(X_train, y_train)
acc_3, precision_3, recall_3 = ngram_3.evaluate(X_test, y_test)


print(f"2-gram - Accuracy: {acc_2}, Precision: {precision_2}, Recall: {recall_2}")
print(f"3-gram - Accuracy: {acc_3}, Precision: {precision_3}, Recall: {recall_3}")

if acc_2 > acc_3:
    print("مدل 2-gram دقت بهتری دارد.")
else:
    print("مدل 3-gram دقت بهتری دارد.")


2-gram - Accuracy: 0.7274119448698315, Precision: 0.8018301562397242, Recall: 0.7274119448698315
3-gram - Accuracy: 0.7289433384379785, Precision: 0.8026407717659014, Recall: 0.7289433384379785
مدل 3-gram دقت بهتری دارد.


<div style="direction: rtl; text-align: right;">
    این کد برای طیقه بندی با استفاده از ngram ها نوشته شده است و از روش multinomial naive bayes استفاده میشود.
    <br><br>
    ابتدا کتابخانه‌های لازم از جمله pandas برای مدیریت داده‌ها، sklearn برای الگوریتم‌های یادگیری ماشین، و tokenizers برای توکن‌سازی متن وارد می‌شود.
    <br><br>
    سپس کلاس NgramClassifier با پارامترهای n و tokenizer_path تعریف شده است.
    <br><br>
    در مرحله بعد و در متد init توکنایزر از فایل مربوطه بارگذاری می‌شود.
    <br><br>
    یک شیء TfidfVectorizer برای تبدیل متن به ویژگی‌های عددی با استفاده از TF-IDF (Term Frequency-Inverse Document Frequency) و ن-گرم‌ها ساخته می‌شود.
    <br><br>
    مدل MultinomialNB برای طبقه‌بندی ایجاد می‌شود.
    <br><br>
    سپس در متد tokenize_text متن ورودی توکنایز شده و به یک رشته تبدیل میشود و در متد train مدل آموزش میبیند، در این متد متن ها توکن سازی شده و به ویژگی های عددی تبدیل میشوند و سپس مدل روی داده های آموزشی فیت میشود.
    <br><br>
    متد evaluate نیز برای ارزیابی مدل استفاده می شود و متون تست توکن سازی شده و به ویژگی های عددی تبدیل  میشوند سپس پیش بینی انجام شده و معیارهای accuracy, precision و recall محاسبه میشوند.
    <br><br>
    در مرحله بعد داده‌ها از یک فایل CSV بارگذاری می‌شود و ردیف‌های خالی در ستون‌های Text و Suggestion حذف می‌شوند. و داده‌ها به دو بخش آموزشی و تست تقسیم می‌شوند (با نسبت 80-20).
    <br><br>
    سپس یک شیء از NgramClassifier برای ن-گرم 2 و 3 ایجاد می‌شود.
    <br><br>
    هر دو مدل با داده‌های آموزشی فیت می‌شوند و سپس ارزیابی می‌شوند و معیار های ارزیابی محاسبه میشوند.
    <br><br>
    مقادیر به دست آمده نشان دهنده این است که:
    <br><br>
    برای مدل 2-gram:
    <br><br>
    Accurscy برابر 0.727 است که هرچه به ۱ نزدیک تر باشد نشان از دقت بالاتر دارد و این مورد خوب است.
    <br><br>
    Precision برابر 0.801 است که نشان میدهد ۸۰ درصد مقادیر مثبتی که پیش بینی شده واقعا مثبت بودند و هر جه این معیار به ۱ نزدیک تر باشد بهتر است.
    <br><br>
    Recall برابر 0.727 است که به معنای این است که تمام نمونه‌های مثبت واقعی به درستی شناسایی شده‌اند. هر چه این مقدار به ۱ نزدیک تر باشد بهتر است.
    <br><br>
    ------------------------------
    <br><br>
    برای مدل 3-gram:
    <br><br>
    Accurscy برابر 0.728 است که هرچه به ۱ نزدیک تر باشد نشان از دقت بالاتر دارد و این مورد خوب است.
    <br><br>
    Precision برابر 0.802 است که نشان میدهد ۸۰ درصد مقادیر مثبتی که پیش بینی شده واقعا مثبت بودند و هر جه این معیار به ۱ نزدیک تر باشد بهتر است.
    <br><br>
    Recall برابر 0.728 است که به معنای این است که تمام نمونه‌های مثبت واقعی به درستی شناسایی شده‌اند. هر چه این مقدار به ۱ نزدیک تر باشد بهتر است.
    <br><br>
    در کل این مقادیر برای مدل 3-gram بیشتر است و نشان دهنده عملکرد بهتر مدل 3-gram دارد.
<div>

# **سوال چهارم:**

*بخش اول:*
---
# پاسخ خود را اینجا بنویسید:

In [23]:
#write your code here


*بخش دوم:*
---
# پاسخ خود را اینجا بنویسید:

In [24]:
#write your code here


# **نکات مهم**

<div dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
    <p><strong>مهلت تحویل بدون جریمه:</strong> ۲۷ مهر ۱۴۰۳</p>
    <p><strong>مهلت تحویل با تاخیر (با جریمه):</strong> ۴ آبان ۱۴۰۳</p>
</div>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">فایل ارسالی شما باید با فرمت زیر نامگذاری شود: <code>NLP_CA1_LASTNAME_STUDENTID.ipynb</code></h4>

<h4 dir="rtl" style="font-family: Vazir; width: 85%;">نحوه انجام این تمرین:</h4>
<ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>برخی سوالات نیاز به نوشتن کد پایتون و محاسبه نتایج دارند، و بقیه آنها دارای پاسخ‌های نوشتاری هستند. برای مسائل کدنویسی، باید تمام بلوک‌های کدی که با <code>#WRITE YOUR CODE HERE</code> مشخص شده‌اند را تکمیل کنید.</li> <li>برای پاسخ‌های متنی، باید متنی که می‌گوید "پاسخ خود را اینجا بنویسید..." را با پاسخ واقعی خود جایگزین کنید.</li> </ul>

<h4 dir="rtl" style="font-family: Vazir; width: 85%;">صداقت علمی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>ما نوت‌بوک‌های تعداد مشخصی از دانشجویان که به صورت تصادفی انتخاب می‌شوند، بررسی خواهیم کرد. این بررسی‌ها اطمینان حاصل می‌کنند که کدی که نوشتید واقعاً پاسخ‌های موجود در نوت‌بوک شما را تولید می‌کند. اگر پاسخ‌های صحیح را در نوت‌بوک خود بدون کدی که واقعاً آن پاسخ‌ها را تولید کند تحویل دهید، این یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> <li>ما همچنین بررسی‌های خودکاری را برای تشخیص سرقت علمی در نوت‌بوک‌های کولب انجام خواهیم داد. کپی کردن کد از دیگران نیز یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> </ul>

<h4 dir="rtl" style="font-family: Vazir; width: 85%;">توضیحات تکمیلی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
<li>
خوانایی و دقت بررسی‌ها در گزارش نهایی از اهمیت ویژه‌ای برخوردار است. به تمرین‌هایی که به صورت کاغذی تحویل داده شوند یا به صورت عکس در سایت بارگذاری شوند، ترتیب اثری داده نخواهد شد.</li>
<li>
 همه‌ی کدهای پیوست گزارش بایستی قابلیت اجرای مجدد داشته باشند. در صورتی که برای اجرا مجدد آن‌ها نیاز به تنظیمات خاصی می‌باشد، بایستی تنظیمات مورد نیاز را نیز در گزارش خود ذکر کنید.  دقت کنید که  تمامی کدها باید توسط شما اجرا شده باشند و نتایج اجرا در فایل کدهای ارسالی مشخص باشد. به کدهایی که نتایج اجرای آن‌ها در فایل ارسالی مشخص نباشد نمره‌ای تعلق نمی‌گیرد.
</li>
<li>
تمرین تا یک هفته بعد از مهلت تعیین شده با تاخیر تحویل گرفته می‌شود. دقت کنید که شما جمعاً برای تمام تکالیف، ۱۴ روز زمان تحویل بدون جریمه دارید که تنها از ۷ روز آن برای هر تمرین می‌توانید استفاده کنید. در صورتی که این ۱۴ روز به اتمام رسیده باشد، به ازای هر روز تاخیر ده درصد جریمه می‌شود.
</li>
<li>توجه کنید این تمرین باید به صورت تک‌نفره انجام شود و پاسخ‌های ارائه شده باید نتیجه فعالیت فرد نویسنده باشد (همفکری و به اتفاق هم نوشتن تمرین نیز ممنوع است). در صورت مشاهده
 تشابه به همه افراد مشارکت‌کننده، نمره تمرین صفر و به استاد گزارش می‌گردد.
 </li>
 <li>برای مطالعه بیشتر درباره‌ی فرمت مارک‌دون می‌توانید از <a href="https://github.com/tajaddini/Persian-Markdown/blob/master/learn-MD.md">این لینک</a> مطالعه کنید.
 </li>

 </ul>
 </div>